File: cpp/src/client/trace_hook.cpp
// 消息轨迹钩子实现（对应 org.apache.rocketmq.client.trace.hook 包）。
//
// 三条硬性约定见 trace_hook.h 头注释：轨迹消息不再被追踪、是否落轨迹由 broker / 消息属性
// 决定、消费侧 msgId 用 offset 侧 ID（对齐 SendResult.offsetMsgId，而不是 SendResult.msgId /
// UNIQ_KEY）。
#include "rocketmq/client/trace_hook.h"

#include "rocketmq/common/logging.h"
#include "rocketmq/common/message_const.h"
#include "rocketmq/common/namespace_util.h"
#include "rocketmq/common/util_all.h"

namespace rocketmq {

namespace {

// ConsumeReturnType 名字 -> ordinal（顺序即 Java ConsumeReturnType 的 ordinal，
// 轨迹 SubAfter 的 contextCode 用的正是它）。未知名字退化为 SUCCESS(0)。
int32_t consumeReturnTypeOrdinal(const std::string& name) {
    if (name == "SUCCESS") return 0;
    if (name == "TIME_OUT") return 1;
    if (name == "EXCEPTION") return 2;
    if (name == "RETURNNULL") return 3;
    if (name == "FAILED") return 4;
    return 0;
}

bool startsWith(const std::string& s, const std::string& prefix) {
    if (prefix.size() > s.size()) return false;
    return s.compare(0, prefix.size(), prefix) == 0;
}

}  // namespace

// ---------------------------------------------------------------- SendMessage
void SendMessageTraceHook::sendMessageBefore(SendMessageContext& context) {
    if (dispatcher_ == nullptr || context.message == nullptr) return;
    const std::string& topic = context.message->topic;
    if (startsWith(topic, dispatcher_->getTraceTopicName())) return;  // 轨迹消息不再被追踪

    auto traceContext = std::make_shared<TraceContext>();
    context.mqTraceContext = traceContext;
    traceContext->traceType = TraceType::PUB;
    traceContext->groupName = NamespaceUtil::withoutNamespace(context.producerGroup);

    TraceBean bean;
    bean.topic = NamespaceUtil::withoutNamespace(topic);
    bean.tags = context.message->getTags();
    bean.keys = context.message->getKeys();
    bean.storeHost = context.brokerAddr;
    bean.bodyLength = static_cast<int32_t>(context.message->body.size());
    bean.msgType = context.msgType;
    traceContext->traceBeans = {bean};
}

void SendMessageTraceHook::sendMessageAfter(SendMessageContext& context) {
    if (dispatcher_ == nullptr || context.message == nullptr) return;
    const std::string& topic = context.message->topic;
    if (startsWith(topic, dispatcher_->getTraceTopicName())) return;
    if (context.mqTraceContext == nullptr) return;
    if (context.sendResult == nullptr) return;
    const SendResult& result = *context.sendResult;
    // broker 侧 traceOn=false 或没带回 region 时不落库（对齐 Java）
    if (result.regionId.empty() || !result.traceOn) return;

    TraceContext& traceContext = *context.mqTraceContext;
    if (traceContext.traceBeans.empty()) return;

    int64_t now = UtilAll::currentTimeMillis();
    int32_t costTime = static_cast<int32_t>((now - traceContext.timeStamp)
                                            / static_cast<int64_t>(traceContext.traceBeans.size()));
    traceContext.costTime = costTime;
    traceContext.isSuccess = (result.sendStatus == SendStatus::SEND_OK);
    traceContext.regionId = result.regionId;

    TraceBean& bean = traceContext.traceBeans[0];
    bean.msgId = result.msgId;               // 客户端 UNIQ_KEY
    bean.offsetMsgId = result.offsetMsgId;   // broker 的 offset 消息 ID
    bean.storeTime = traceContext.timeStamp + static_cast<int64_t>(costTime) / 2;
    dispatcher_->append(context.mqTraceContext);
}

// ---------------------------------------------------------------- ConsumeMessage
void ConsumeMessageTraceHook::consumeMessageBefore(ConsumeMessageContext& context) {
    if (dispatcher_ == nullptr || context.msgList.empty()) return;

    auto traceContext = std::make_shared<TraceContext>();
    context.mqTraceContext = traceContext;
    traceContext->traceType = TraceType::SUB_BEFORE;
    traceContext->groupName = NamespaceUtil::withoutNamespace(context.consumerGroup);

    std::vector<TraceBean> beans;
    for (const MessageExt& msg : context.msgList) {
        // Java 只跳过 null（C++ 的 vector 里没有 null，故不额外过滤），
        // 也不按空 topic 跳过 —— 与 Python 参考实现保持一致。
        std::string traceOn = msg.getProperty(TraceConstants::PROPERTY_TRACE_SWITCH);
        if (traceOn == "false") continue;  // 消息级关闭轨迹
        TraceBean bean;
        bean.topic = NamespaceUtil::withoutNamespace(msg.topic);
        bean.msgId = msg.msgId;            // 消费侧用 offset 侧 ID（对齐 SendResult.offsetMsgId）
        bean.tags = msg.getTags();
        bean.keys = msg.getKeys();
        bean.storeTime = msg.storeTimestamp;
        bean.bodyLength = msg.storeSize;
        bean.retryTimes = msg.reconsumeTimes;
        std::string region = msg.getProperty(MessageConst::PROPERTY_MSG_REGION);
        traceContext->regionId = region;   // 逐条覆盖（最后一条生效，对齐 Python/Java）
        beans.push_back(bean);
    }
    if (!beans.empty()) {
        traceContext->traceBeans = beans;
        traceContext->timeStamp = UtilAll::currentTimeMillis();
        dispatcher_->append(traceContext);
    }
}

void ConsumeMessageTraceHook::consumeMessageAfter(ConsumeMessageContext& context) {
    if (dispatcher_ == nullptr || context.msgList.empty()) return;
    if (context.mqTraceContext == nullptr) return;
    if (context.mqTraceContext->traceBeans.empty()) return;

    TraceContext& subBefore = *context.mqTraceContext;

    auto subAfter = std::make_shared<TraceContext>();
    subAfter->traceType = TraceType::SUB_AFTER;
    subAfter->regionId = subBefore.regionId;
    subAfter->groupName = NamespaceUtil::withoutNamespace(subBefore.groupName);
    subAfter->requestId = subBefore.requestId;        // 与 SubBefore 共用一个 request_id
    subAfter->accessChannel = context.accessChannel;
    subAfter->isSuccess = context.success;
    int64_t now = UtilAll::currentTimeMillis();
    int32_t costTime = static_cast<int32_t>((now - subBefore.timeStamp)
                                            / static_cast<int64_t>(context.msgList.size()));
    subAfter->costTime = costTime;
    subAfter->traceBeans = subBefore.traceBeans;      // 同一批消息，msgId 沿用 offset 侧 ID

    // props 里放的是 ConsumeReturnType 的**名字**（Java 按名查 ordinal），未知退化为 SUCCESS。
    subAfter->contextCode = consumeReturnTypeOrdinal(context.props);

    dispatcher_->append(subAfter);
}

// ---------------------------------------------------------------- EndTransaction
void EndTransactionTraceHook::endTransaction(EndTransactionContext& context) {
    if (dispatcher_ == nullptr || context.message == nullptr) return;
    const std::string& topic = context.message->topic;
    if (startsWith(topic, dispatcher_->getTraceTopicName())) return;  // 轨迹消息不再被追踪

    auto traceContext = std::make_shared<TraceContext>();
    traceContext->traceType = TraceType::END_TRANSACTION;
    traceContext->groupName = NamespaceUtil::withoutNamespace(context.producerGroup);

    TraceBean bean;
    bean.topic = NamespaceUtil::withoutNamespace(topic);
    bean.tags = context.message->getTags();
    bean.keys = context.message->getKeys();
    bean.storeHost = context.brokerAddr;
    bean.msgType = static_cast<int32_t>(TraceMessageType::TRANS_COMMIT);
    bean.clientHost = dispatcher_->clientId();
    bean.msgId = context.msgId;
    bean.transactionState = context.transactionState;
    bean.transactionId = context.transactionId;
    bean.fromTransactionCheck = context.fromTransactionCheck;
    traceContext->traceBeans = {bean};
    traceContext->timeStamp = UtilAll::currentTimeMillis();

    std::string region = context.message->getProperty(MessageConst::PROPERTY_MSG_REGION);
    traceContext->regionId = region.empty() ? std::string(TraceConstants::DEFAULT_TRACE_REGION_ID)
                                            : region;
    dispatcher_->append(traceContext);
}

}  // namespace rocketmq

File: dotnet/src/RocketMQ.Client/Client/TraceHook.cs
// 消息轨迹钩子（对应 org.apache.rocketmq.client.trace.hook 包）。
//
//   * SendMessageTraceHook    ← SendMessageTraceHookImpl
//   * ConsumeMessageTraceHook ← ConsumeMessageTraceHookImpl
//   * EndTransactionTraceHook ← EndTransactionTraceHookImpl
//
// 两条硬性约定：
//   1. 轨迹消息本身不再被追踪：before/after 都先看 topic 是否以轨迹 topic 开头，是则直接
//      return（否则轨迹会自我复制）。
//   2. 是否落轨迹由 broker 说了算：发送侧看 SendResult 的 RegionId / TraceOn（由 SEND 响应头
//      的 MSG_REGION / TRACE_ON 解析而来，broker 默认 traceOn=true）；消费侧看消息属性
//      TRACE_ON 是否为 "false"。
using RocketMQ.Common;

namespace RocketMQ.Client;

/// <summary>发送侧轨迹钩子（对应 SendMessageTraceHookImpl）。</summary>
public sealed class SendMessageTraceHook : ISendMessageHook
{
    private readonly AsyncTraceDispatcher _localDispatcher;

    public SendMessageTraceHook(AsyncTraceDispatcher localDispatcher) => _localDispatcher = localDispatcher;

    public string HookName() => "SendMessageTraceHook";

    public void SendMessageBefore(SendMessageContext context)
    {
        if (context is null || context.Message is null)
        {
            return;
        }

        string topic = context.Message.Topic ?? string.Empty;
        if (topic.StartsWith(_localDispatcher.GetTraceTopicName(), StringComparison.Ordinal))
        {
            return;
        }

        var traceContext = new TraceContext
        {
            TraceType = RocketMQ.Client.TraceType.Pub,
            GroupName = NamespaceUtil.WithoutNamespace(context.ProducerGroup),
        };
        context.MqTraceContext = traceContext;
        var bean = new TraceBean
        {
            Topic = NamespaceUtil.WithoutNamespace(topic),
            Tags = context.Message.Tags ?? string.Empty,
            Keys = context.Message.Keys ?? string.Empty,
            StoreHost = context.BrokerAddr ?? string.Empty,
            BodyLength = context.Message.Body?.Length ?? 0,
            MsgType = context.MsgType,
        };
        traceContext.TraceBeans = new List<TraceBean> { bean };
    }

    public void SendMessageAfter(SendMessageContext context)
    {
        if (context is null || context.Message is null || context.SendResult is null)
        {
            return;
        }

        string topic = context.Message.Topic ?? string.Empty;
        if (topic.StartsWith(_localDispatcher.GetTraceTopicName(), StringComparison.Ordinal))
        {
            return;
        }

        if (context.MqTraceContext is not TraceContext traceContext || traceContext.TraceBeans.Count == 0)
        {
            return;
        }

        SendResult result = context.SendResult;
        // broker 侧 traceOn=false 或没带回 region 时不落库（对齐 Java）。
        if (string.IsNullOrEmpty(result.RegionId) || !result.TraceOn)
        {
            return;
        }

        TraceBean bean = traceContext.TraceBeans[0];
        int costTime = (int)((UtilAll.CurrentTimeMillis() - traceContext.TimeStamp) / traceContext.TraceBeans.Count);
        traceContext.CostTime = costTime;
        traceContext.IsSuccess = result.SendStatus == SendStatus.SendOk;
        traceContext.RegionId = result.RegionId;
        bean.MsgId = result.MsgId ?? string.Empty;
        bean.OffsetMsgId = result.OffsetMsgId ?? string.Empty;
        bean.StoreTime = traceContext.TimeStamp + (costTime / 2);
        _localDispatcher.Append(traceContext);
    }
}

/// <summary>消费侧轨迹钩子（对应 ConsumeMessageTraceHookImpl，SubBefore / SubAfter 共用 request_id）。</summary>
public sealed class ConsumeMessageTraceHook : IConsumeMessageHook
{
    private readonly AsyncTraceDispatcher _localDispatcher;

    public ConsumeMessageTraceHook(AsyncTraceDispatcher localDispatcher) => _localDispatcher = localDispatcher;

    public string HookName() => "ConsumeMessageTraceHook";

    public void ConsumeMessageBefore(ConsumeMessageContext context)
    {
        if (context is null || context.MsgList.Count == 0)
        {
            return;
        }

        var traceContext = new TraceContext
        {
            TraceType = RocketMQ.Client.TraceType.SubBefore,
            GroupName = NamespaceUtil.WithoutNamespace(context.ConsumerGroup),
        };
        context.MqTraceContext = traceContext;
        var beans = new List<TraceBean>();
        foreach (MessageExt msg in context.MsgList)
        {
            if (msg is null)
            {
                continue;
            }

            string? traceOn = msg.GetProperty(MessageConst.PropertyTraceSwitch);
            if (traceOn is not null && traceOn == "false")
            {
                continue;
            }

            string regionId = msg.GetProperty(MessageConst.PropertyMsgRegion);
            var bean = new TraceBean
            {
                Topic = NamespaceUtil.WithoutNamespace(msg.Topic),
                MsgId = msg.MsgId ?? string.Empty,
                Tags = msg.Tags ?? string.Empty,
                Keys = msg.Keys ?? string.Empty,
                StoreTime = msg.StoreTimestamp,
                BodyLength = msg.StoreSize,
                RetryTimes = msg.ReconsumeTimes,
            };
            traceContext.RegionId = regionId ?? string.Empty;
            beans.Add(bean);
        }

        if (beans.Count > 0)
        {
            traceContext.TraceBeans = beans;
            traceContext.TimeStamp = UtilAll.CurrentTimeMillis();
            _localDispatcher.Append(traceContext);
        }
    }

    public void ConsumeMessageAfter(ConsumeMessageContext context)
    {
        if (context is null || context.MsgList.Count == 0)
        {
            return;
        }

        if (context.MqTraceContext is not TraceContext subBefore || subBefore.TraceBeans.Count == 0)
        {
            return;
        }

        var subAfter = new TraceContext
        {
            TraceType = RocketMQ.Client.TraceType.SubAfter,
            RegionId = subBefore.RegionId,
            GroupName = NamespaceUtil.WithoutNamespace(subBefore.GroupName),
            RequestId = subBefore.RequestId,
            AccessChannel = context.AccessChannel,
            IsSuccess = context.Success,
            TraceBeans = subBefore.TraceBeans,
        };
        subAfter.CostTime = (int)((UtilAll.CurrentTimeMillis() - subBefore.TimeStamp) / context.MsgList.Count);

        // ConsumeContextType 是 ConsumeReturnType 的**名字**，其 ordinal 即轨迹里的 contextCode。
        if (context.Props is not null && context.Props.TryGetValue("ConsumeContextType", out string? ct) && ct is not null)
        {
            subAfter.ContextCode = ReturnTypeOrdinal(ct);
        }

        _localDispatcher.Append(subAfter);
    }

    /// <summary>ConsumeReturnType 名字 → ordinal（对齐 Java org.apache.rocketmq.client.consumer.listener.
    /// ConsumeReturnType：SUCCESS=0、TIME_OUT=1、EXCEPTION=2、RETURNNULL=3、FAILED=4）。
    /// ⚠ 这里**不是** ConsumeConcurrentlyStatus（CONSUME_SUCCESS/RECONSUME_LATER），写错会让
    /// 控制台的 contextCode 整体错位。</summary>
    private static int ReturnTypeOrdinal(string name) => name switch
    {
        "SUCCESS" => 0,
        "TIME_OUT" => 1,
        "EXCEPTION" => 2,
        "RETURNNULL" => 3,
        "FAILED" => 4,
        _ => 0,
    };
}

/// <summary>事务收尾轨迹钩子（对应 EndTransactionTraceHookImpl）。</summary>
public sealed class EndTransactionTraceHook : IEndTransactionHook
{
    private readonly AsyncTraceDispatcher _localDispatcher;

    public EndTransactionTraceHook(AsyncTraceDispatcher localDispatcher) => _localDispatcher = localDispatcher;

    public string HookName() => "EndTransactionTraceHook";

    public void EndTransaction(EndTransactionContext context)
    {
        if (context is null || context.Message is null)
        {
            return;
        }

        string topic = context.Message.Topic ?? string.Empty;
        if (topic.StartsWith(_localDispatcher.GetTraceTopicName(), StringComparison.Ordinal))
        {
            return;
        }

        Message msg = context.Message;
        var tuxeContext = new TraceContext
        {
            TraceType = RocketMQ.Client.TraceType.EndTransaction,
            GroupName = NamespaceUtil.WithoutNamespace(context.ProducerGroup),
        };
        var bean = new TraceBean
        {
            Topic = NamespaceUtil.WithoutNamespace(topic),
            Tags = msg.Tags ?? string.Empty,
            Keys = msg.Keys ?? string.Empty,
            StoreHost = context.BrokerAddr ?? string.Empty,
            MsgType = TraceMessageType.TransCommit,
            ClientHost = _localDispatcher.ClientId(),
            MsgId = context.MsgId ?? string.Empty,
            TransactionId = context.TransactionId,
            TransactionState = TransactionStateName(context.TransactionState),
            FromTransactionCheck = context.FromTransactionCheck,
        };
        string? regionId = msg.GetProperty(MessageConst.PropertyMsgRegion);
        tuxeContext.RegionId = regionId ?? MixAll.DefaultTraceRegionId;
        tuxeContext.TraceBeans = new List<TraceBean> { bean };
        tuxeContext.TimeStamp = UtilAll.CurrentTimeMillis();
        _localDispatcher.Append(tuxeContext);
    }

    /// <summary>把 LocalTransactionState 映射成 Java 枚举名（Java 写的是 name()，即
    /// COMMIT_MESSAGE / ROLLBACK_MESSAGE / UNKNOW）。⚠ .NET 枚举成员是 CommitMessage 这种
    /// 帕斯卡命名，直接 ToString() 会写出 "CommitMessage"，控制台认不出来。</summary>
    private static string TransactionStateName(object? state) => state switch
    {
        null => string.Empty,
        string s => s,
        LocalTransactionState st => st switch
        {
            LocalTransactionState.CommitMessage => "COMMIT_MESSAGE",
            LocalTransactionState.RollbackMessage => "ROLLBACK_MESSAGE",
            LocalTransactionState.Unknow => "UNKNOW",
            _ => st.ToString(),
        },
        _ => state.ToString() ?? string.Empty,
    };
}
